### PR TITLE
Upgrade to sundrio 0.40.0

### PR DESCRIPTION
**Pull request by iocanel**
-------
## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by manusa** [comment]
image
Cool!

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Comment by manusa** [comment]
Fixes #3102

**Review by metacosm — CHANGES_REQUESTED** [review]

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java [hunk 1]
```diff
@@ -210,11 +208,13 @@ void jokerequestCRDShouldWork() {
       final List<CustomResourceColumnDefinition> printerColumns = version
         .getAdditionalPrinterColumns();
       assertEquals(2, printerColumns.size());
-      CustomResourceColumnDefinition columnDefinition = printerColumns.get(0);
+      CustomResourceColumnDefinition columnDefinition = printerColumns.stream().filter(p -> p.getJsonPath().equals(".spec.category")).findFirst().orElse(null);
```
[the remark this answers is not in the document]
Comment (iocanel) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:211]: Then the ordering of printer columns is broken.

[the remark this answers is not in the document]
Comment (iocanel) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:211]: On second thought, I don't see how is there such thing as order of printer columns? A printer columns may be defined anywhere in the graph, how can you guarantee the order?

[the remark this answers is not in the document]
Comment (metacosm) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:211]: That was the goal of this PR: https://github.com/fabric8io/kubernetes-client/pull/3071/files

File: crd-generator/api/src/main/java/io/fabric8/crd/generator/v1/CustomResourceHandler.java [hunk 2]
```diff
@@ -82,7 +80,7 @@ protected void addDecorators(CustomResourceInfo config, TypeDef def,
       resources.decorate(new AddLabelSelectorPathDecorator(name, version, path));
     });
 
-    if (Types.findStatusProperty(def).isPresent()) {
+    if (config.specClassName().isPresent()) {
```
Comment (metacosm) [inline crd-generator/api/src/main/java/io/fabric8/crd/generator/v1/CustomResourceHandler.java:83]: Wrong method call: this should be `config.statusClassName()`.

**Review by metacosm — CHANGES_REQUESTED** [review]

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:211]: Cf https://github.com/fabric8io/kubernetes-client/blob/62e2bcd260e7bdc017eb19aec2a8e2adbdbdeba6/crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java#L207

[the remark this answers is not in the document]
Comment (iocanel) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:211]: That pull request may or not work, given that it was applied on a test that was disabled and no other test has been provided.
I am going to revert that change and align the test in order to pass.

[the remark this answers is not in the document]
Comment (iocanel) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java]: Fixed the `SortPrinterColumnDecorator` and reverted that change.

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java [hunk 3]
```diff
@@ -210,11 +208,13 @@ void jokerequestCRDShouldWork() {
...
+      assertNotNull(columnDefinition);
       assertEquals("string", columnDefinition.getType());
       assertEquals(".spec.category", columnDefinition.getJsonPath());
       assertEquals("jokeCategory", columnDefinition.getName());
-      columnDefinition = printerColumns.get(1);
+      columnDefinition = printerColumns.stream().filter(p -> p.getJsonPath().equals(".spec.excluded")).findFirst().orElse(null);
```
[the remark this answers is not in the document]
Comment (iocanel) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java]: Fixed the `SortPrinterColumnDecorator` and reverted that change.

**Comment by iocanel** [comment]
@manusa: Addressed all concerns raised by @metacosm.

**Comment by metacosm** [comment]
Out of curiosity what was wrong with the printer columns sorting?

**Comment by manusa** [comment]
After analyzing the generated code with 0.40.0 and 0.30.6, these are some of the differences:

| 0.40.0 | 0.30.6 | Comments |
| -------- | -------- | --------------- |
| Canonical class references | imports | Improved resiliency |
| build methods use ternary+streams | build methods have if-statements+for-loops | For all cases except for `Set` the resulting collection is added to a new ArrayList. For the `Set` case this might be a problem if the resulting collection is manipulated by the consuming code |
| canVisit template probably has end-of-line spaces | | nit: would be good to remove those spaces |
| Fluent type interface is now less restrictive `Fluent<F>` | Fluent type interface is more restrictive `Fluent<F extends Fluent<F>>` | As agreed with @iocanel this is a bug that needs to be addressed |
| Visitor interface has no longer the `@FunctionalInterface` annotation | | |
| `withNewXXX(StringBuffer)` / `withNewXXX(StringBuilder)` methods lost | | It's very unlikely that any consuming code is making use of this methods, it seems reasonable to keep this breaking change |

**Pull request merged**